flytepropeller/pkg/controller/nodes/common/utils.go

```diff
@@ -67,3 +78,82 @@ func GetTargetEntity(ctx context.Context, nCtx interfaces.NodeExecutionContext)
 	}
 	return targetEntity
 }
+
+// OffloadLargeLiteral offloads the large literal if meets the threshold conditions
+func OffloadLargeLiteral(ctx context.Context, datastore *storage.DataStore, dataReference storage.DataReference,
+	toBeOffloaded *idlcore.Literal, literalOffloadingConfig config.LiteralOffloadingConfig) error {
+	literalSizeBytes := uint64(proto.Size(toBeOffloaded))
+	literalSizeMB := literalSizeBytes / MB
+	// check if the literal is large
+	if literalSizeMB >= literalOffloadingConfig.MaxSizeInMBForOffloading {
+		errString := fmt.Sprintf("Literal size [%d] MB is larger than the max size [%d] MB for offloading", literalSizeMB, literalOffloadingConfig.MaxSizeInMBForOffloading)
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+	if literalSizeMB < literalOffloadingConfig.MinSizeInMBForOffloading {
+		logger.Debugf(ctx, "Literal size [%d] MB is smaller than the min size [%d] MB for offloading", literalSizeMB, literalOffloadingConfig.MinSizeInMBForOffloading)
+		return nil
+	}
+
+	inferredType := validators.LiteralTypeForLiteral(toBeOffloaded)
+	if inferredType == nil {
+		errString := "Failed to determine literal type for offloaded literal"
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+
+	// offload the literal
+	if err := datastore.WriteProtobuf(ctx, dataReference, storage.Options{}, toBeOffloaded); err != nil {
+		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
+		return err
+	}
+
+	// update the literal with the offloaded URI, size and inferred type
+	toBeOffloaded.Value = &idlcore.Literal_OffloadedMetadata{
+		OffloadedMetadata: &idlcore.LiteralOffloadedMetadata{
+			Uri:          dataReference.String(),
+			SizeBytes:    literalSizeBytes,
+			InferredType: inferredType,
+		},
+	}
+	logger.Infof(ctx, "Offloaded literal at location [%s] with size [%d] MB and inferred type [%s]", dataReference, literalSizeMB, inferredType)
+	return nil
+}
+
+// CheckOffloadingCompat checks if the upstream and downstream nodes are compatible with the literal offloading feature and returns an error if not contained in phase info object
+func CheckOffloadingCompat(ctx context.Context, nCtx interfaces.NodeExecutionContext, inputLiterals map[string]*core.Literal, node v1alpha1.ExecutableNode, literalOffloadingConfig config.LiteralOffloadingConfig) *handler.PhaseInfo {
+	consumesOffloadLiteral := false
+	for _, val := range inputLiterals {
+		if val != nil && val.GetOffloadedMetadata() != nil {
+			consumesOffloadLiteral = true
+			break
+		}
+	}
+	var phaseInfo handler.PhaseInfo
+	switch node.GetKind() {
+	case v1alpha1.NodeKindTask:
```

Comment:
should we move the check for the node kind earlier and return if it's not `NodeKindTask`?

Reply:
Done